### ISSUE TITLE
Client hangs on Subscribe(); token.Wait()

### ISSUE THREAD

**Issue by wari** [eclipse/paho.mqtt.golang#232]
Looking at the samples, we are told to do token.Wait() after a Subscribe(), but unfortunately, this can cause a hang on my side. Removing the token.Wait() solves the issue. 

```
	if token := c.Subscribe("go-mqtt/sample", 0, nil); token.Wait() && token.Error() != nil {
		fmt.Println(token.Error())
		os.Exit(1)
	}
```

Here's a debug trace.

* Without token.Wait()

```
2018-07-25T16:07:34+08:00 |INFO| Setting up consumer
2018-07-25T16:07:34+08:00 |INFO| Done
2018-07-25T16:07:34+08:00 |INFO| Launching consumer
2018-07-25T16:07:34+08:00 |DEBU| [[client]   Connect()]
2018-07-25T16:07:34+08:00 |DEBU| [[store]    memorystore initialized]
2018-07-25T16:07:34+08:00 |DEBU| [[client]   about to write new connect msg]
2018-07-25T16:07:34+08:00 |DEBU| [[client]   socket connected to broker]
2018-07-25T16:07:34+08:00 |DEBU| [[client]   Using MQTT 3.1.1 protocol]               
2018-07-25T16:07:34+08:00 |DEBU| [[net]      connect started]
2018-07-25T16:07:34+08:00 |DEBU| [[net]      received connack]
2018-07-25T16:07:34+08:00 |DEBU| [[client]   client is connected]
2018-07-25T16:07:34+08:00 |DEBU| [[client]   exit startClient]
2018-07-25T16:07:34+08:00 |INFO| Done
2018-07-25T16:07:34+08:00 |INFO| Subscribing to: filebeat/lines
2018-07-25T16:07:34+08:00 |DEBU| [[client]   enter Subscribe]
2018-07-25T16:07:34+08:00 |DEBU| [[net]      incoming started]
2018-07-25T16:07:34+08:00 |DEBU| [[net]      logic started]
2018-07-25T16:07:34+08:00 |DEBU| [[net]      outgoing started]
2018-07-25T16:07:34+08:00 |DEBU| [[net]      outgoing waiting for an outbound message]
2018-07-25T16:07:34+08:00 |DEBU| [[client]   SUBSCRIBE: dup: false qos: 1 retain: false rLength: 0 MessageID: 0 topics: [filebeat/lines]]
2018-07-25T16:07:34+08:00 |DEBU| [[net]      Received Message]
2018-07-25T16:07:34+08:00 |DEBU| [[client]   exit Subscribe]
2018-07-25T16:07:34+08:00 |INFO| Done
```

* With token.Wait()

See: https://pastebin.com/R0SaqMeq

**Comment by alsm** [comment]
From the trace you have it looks like you are never getting a suback for the subscribe, what mqtt server are you using?

**Comment by wari** [comment]
The MQTT server I'm using is from Solace PubSub+ Cloud. i.e. https://console.solace.cloud/login

**Comment by alsm** [comment]
It seems from what you say that the issue might be with the MQTT server, the fact that no SUBACK turns up at all in the trace. If possible try and recreate the situation with a different client and see if the same thing happens?

**Comment by wari** [comment]
I did a python program as well as do a packet capture for it as well. If there are messages in the queue, the server will publish messages first on Clean Session = 0. Based on the spec:

> If CleanSession is set to 0, the Server MUST resume communications with the Client based on state from the current Session (as identified by the Client identifier). If there is no Session associated with the Client identifier the Server MUST create a new Session. The Client and Server MUST store the Session after the Client and Server are disconnected.

The packet capture for the python program is as follows:

```
  343 22.023172429 192.168.40.103 → 52.74.21.247 MQTT 140 Connect Command
  345 22.029223967 52.74.21.247 → 192.168.40.103 MQTT 70 Connect Ack
  347 22.029254321 52.74.21.247 → 192.168.40.103 MQTT 156 Publish Message (id=1020) [a/b], Publish Message (id=1021) [a/b], Publish Message (id=1022) [a/b], Publish Message (id=1023) [a/b], Publish Message (id=1024) [a/b], Publish Message (id=1) [a/b], Publish Message (id=2) [a/b], Publish Message (id=3) [a/b], Publish Message (id=4) [a/b]
  349 22.029259824 52.74.21.247 → 192.168.40.103 MQTT 76 Publish Message (id=5) [a/b]
  351 22.029641227 192.168.40.103 → 52.74.21.247 MQTT 76 Subscribe Request (id=1) [a/b]
  352 22.033844440 52.74.21.247 → 192.168.40.103 MQTT 71 Subscribe Ack (id=1)
  353 22.033861139 192.168.40.103 → 52.74.21.247 MQTT 106 Publish Ack (id=1020), Publish Ack (id=1021), Publish Ack (id=1022), Publish Ack (id=1023), Publish Ack (id=1024), Publish Ack (id=1), Publish Ack (id=2), Publish Ack (id=3), Publish Ack (id=4), Publish Ack (id=5)
```

and the output of the python `paho-mqtt` program:

```
Connected with result code 0
Subscribing
Done Subscribe
a/b b'A'
a/b b'A'
a/b b'A'
a/b b'A'
a/b b'A'
a/b b'A'
a/b b'A'
a/b b'A'
a/b b'A'
a/b b'A'
############## Client ON Subscribe
<paho.mqtt.client.Client object at 0x7f1cc2e98860> None 1 (1,)
```

So I decided to use the `token.Wait()` to see what conditions would hang my program:

```
  342 2.200720148 192.168.40.103 → 52.74.21.247 MQTT 87 Subscribe Request (id=1) [filebeat/lines]
  675 2.459266810 52.74.21.247 → 192.168.40.103 MQTT 2962 Publish Message (id=263) [filebeat/lines], Subscribe Ack (id=1), Publish Message (id=264) [filebeat/lines]
```

The `Subscribe Ack` came in after 260 1K-ish messages later which is about 250 ms time span.

I'm not sure if the SubAck sandwiched in between a Publish caused the hang though.

**Comment by wari** [comment]
BTW, do note that on the same queue with paho-mqtt-python, the API processes all the messages without the hang, even though it got the suback long after processing a lot of messages.

**Comment by alsm** [comment]
The messagehandler that you set on your subscribe, is it getting invoked at all for the incoming messages?

**Comment by wari** [comment]
The message handler did not get invoked at all. If it does, then there should be no problems I believe.

I'm using the DefaultMessageHandler though, and did not specify one on the subscribe call.

**Comment by alsm** [comment]
It sounds/looks like something is blocking the router, and the only synchronous call in there is to the messagehandler. Can you try setting SetOrderMatters(false) on your client options, that will cause the mesagehandler to be called async. It's not the solution but it'll let us know if that's where it's getting blocked.

**Comment by wari** [comment]
Yup that works, I can see my PublishHandler being called, and after the suback, the "Consumer Subscribe Done" message appeared.

```
2018-07-27T10:36:10+08:00 |DEBU| In PublishHandler
2018-07-27T10:36:10+08:00 |DEBU| [[net]      logic waiting for msg on ibound]
2018-07-27T10:36:10+08:00 |DEBU| [[net]      logic got msg on ibound]
2018-07-27T10:36:10+08:00 |DEBU| [[net]      received suback, id: 1]
2018-07-27T10:36:10+08:00 |DEBU| [[net]      obound priority msg to write, type *packets.PubackPacket]                                                                     
2018-07-27T10:36:10+08:00 |DEBU| [[net]      granted qoss [1]]
2018-07-27T10:36:10+08:00 |DEBU| [[net]      outgoing waiting for an outbound message]
2018-07-27T10:36:10+08:00 |DEBU| [[net]      logic waiting for msg on ibound]                                                                                              
2018-07-27T10:36:10+08:00 |DEBU| [[net]      Received Message]                                                                                                             
2018-07-27T10:36:10+08:00 |DEBU| [[net]      Received Message]                                                                                                             
2018-07-27T10:36:10+08:00 |DEBU| [[net]      logic got msg on ibound]                                                                                                      
2018-07-27T10:36:10+08:00 |DEBU| [[net]      received publish, msgId: 462]
2018-07-27T10:36:10+08:00 |DEBU| [[net]      putting msg on onPubChan]                                                                                                     
2018-07-27T10:36:10+08:00 |DEBU| [[net]      done putting msg on incomingPubChan]
2018-07-27T10:36:10+08:00 |INFO| Consumer Subscribe Done
```

**Comment by wari** [comment]
I now know what is happens. See https://github.com/eclipse/paho.mqtt.golang/blob/88c4622b8e24c52f64a0caaa28e40b91629bb6e6/cmd/sample/main.go#L121-L125 where the choke is processed after the subscribe call. If you `go func() for{}`  this bit and put it before the subscription call, `choke` will be processed immediately, and the net effect is that messages will not get blocked by the choke.

This does not happen when there are about 100 messages in the queue, but if there are more, the choke block will prevent messages from being processed (none will be acked after about 200 messages and suback comes in later) At this point the whole system is blocked. The Solace broker message window size by default is 255, and it will try to fill that window on the first "publish".

When the block happens, those messages that are pubacked will be "lost".

**Comment by wari** [comment]
The other APIs I've worked with (not MQTT) will grab the window size worth of messages, and set it to zero to inform the broker that it cannot process any more messages yet until some are consumed.

Anyway, current workaround is to 

1. Set a buffered channel for the choke (which then gives you a different set of problems), or
2. Set the actual publish handler (in this case is the for loop here) to start earlier.

**Comment by wari** [comment]
I slowed down my publishHandler (`time.Sleep()`), I noticed a pattern, only 100 messages are puback before paho process the next message, and the Solace broker does report 255 messages sent and 100 is being ack'ed.

So, in order to fix this issue, due to the fact that the window size in Solace is 255, and noting that 100 messages are ack'ed at once, I did a grep for 100, and changed https://github.com/eclipse/paho.mqtt.golang/blob/master/options.go#L112 to 255 and the problem goes away immediately.

So `opts.SetMessageChannelDepth(255)` will work, took me a while to figure out, but I believe that the API should not hang when the suback happens only after a couple of hundred messages away.
